src/stir_ldt/prover.rs

```diff
@@ -288,9 +209,128 @@ where
+            .evaluate_over_domain_by_ref(g_domain.backing_domain)
+            .evals;
+
+        // TODO: `stack_evaluations` and `restructure_evaluations` are really in-place algorithms.
+        // They also partially overlap and undo one another. We should merge them.
+        let g_evals_folded = utils::stack_evaluations(g_evals, self.0.folding_factor);
+
+        // TODO: if fold_type is always Naive, this is just an assertion, nothing happens here
+        //
+        // At this point folded evals is a matrix of size (new_domain.size()) X (1 << folding_factor)
+        // This allows for the evaluation of the virutal function using an interpolation on the rows.
+        // TODO: for stir we do only Naive, so this will need to be adapted.
+        let g_evals_folded = restructure_evaluations(
+            g_evals_folded,
+            FoldType::Naive,
+            g_domain.backing_domain.group_gen(),
+            g_domain.backing_domain.group_gen_inv(),
+            self.0.folding_factor,
+        );
+
+        // The leaves of the merkle tree are the k points that are the
+        // roots of unity of the index in the previous domain. This
+        // allows for the evaluation of the virtual function.
+        #[cfg(not(feature = "parallel"))]
+        let leaf_iterator = g_evals_folded.chunks_exact(1 << self.0.folding_factor);
+
+        #[cfg(feature = "parallel")]
+        let leaf_iterator = g_evals_folded.par_chunks_exact(1 << self.0.folding_factor);
+
+        let g_merkle = MerkleTree::<MerkleConfig>::new(
+            &self.0.leaf_hash_params,
+            &self.0.two_to_one_params,
+            leaf_iterator,
+        )
+        .unwrap();
+
+        (g_domain, g_evals_folded, g_merkle)
+    }
+
+    fn stir_phase(
+        &self,
+        merlin: &mut Merlin,
+        num_queries: usize,
+        ctx: &mut RoundContext<F, MerkleConfig>,
+    ) -> ProofResult<Vec<usize>> {
+        let stir_gen = &mut ChaCha20Rng::from_seed(merlin.challenge_bytes()?);
+        let size_of_folded_domain = ctx.f_domain.folded_size(self.0.folding_factor);
+        // Obtain t random integers between 0 and size of the folded domain.
+        // These are the r_shifts from the paper.
+
+        // TODO: this could return fewer than `round_parameters.num_queries` elements
+        let r_shift_indexes =
+            utils::dedup((0..num_queries).map(|_| stir_gen.gen_range(0..size_of_folded_domain)));
+
+        let virtual_evals = self.indexes_to_coset_evaluations(
+            r_shift_indexes.clone(),
+            1 << self.0.folding_factor,
+            &ctx.evals,
+        );
+
+        // Merkle proof for the previous evaluations.
+        let merkle_proof = ctx
+            .merkle
+            .generate_multi_proof(r_shift_indexes.clone())
+            .unwrap();
+
+        ctx.merkle_proofs
+            .push((merkle_proof, virtual_evals.clone()));
+
+        if ctx.round_1 {
+            match self.0.fold_optimization {
```

Comment:
Why are we doing this here? We should just commit to the interpolated values (as in the WHIR impl)